rfd/0105-headless-authentication.md

```diff
@@ -0,0 +1,406 @@
+
+#### `rpc GetHeadlessAuthentication`
+
+This endpoint is used by Teleport clients to retrieve headless login request details before prompting the user for approval/denial.
+
+It is authorized for roles with `read` permissions on the `headless_authentication` resource kind, or for the user who requested headless authentication.
+
+#### `rpc UpdateHeadlessAuthenticationState`
+
+This endpoint is used by Teleport clients to update headless login request state to approved or denied. 
+
+It is only authorized authorized for the user who requested headless authentication. Additionally, when updating the state to `APPROVED`, the client must provide a valid MFA challenge response for the user. An MFA challenge can be requested from the existing rpc `CreateAuthenticateChallenge`.
+
+#### `rpc GenerateUserCerts`
+
+This is an existing endpoint used to generate user certs. We will add `headless_authentication_id` field to retrieve request details from the headless auth request saved on the Auth server. We will add the `headless_authentication_token` field to authorize the certificate generation for the headless authentication.
+
+```proto
+// UserCertRequest specifies certificate-generation parameters
+// for a user.
+message UserCertsRequest {
+  ...
+  // HeadlessAuthenticationID is the id of a headless authentication entity.
+  // When set, the actual cert request will be formed from the headless authentication
+  // details saved on the Auth server.
+  string headless_authentication_id = 17;
+
+  // HeadlessAuthenticationToken is a secret token used to authorize generating
+  // certificates for a successful headless authentication. This is only known by 
+  // the Proxy Server performing the headless authentication flow and the Auth server.
+  string headless_authentication_token = 18;
+}
+```
+
+### UX
+
+#### `tsh --headless`
+
+We will add a new `--headless` flag to `tsh` which can be used to authenticate for a single `tsh` request. When this flag is provided, `tsh` will prompt the user to complete headless authentication on their local machine from the URL `https://proxy.example.com/headless/<id>/approve`. Once the user completes local authentication, `tsh` will receive credentials to complete the request.
+
+```console
+$ tsh --headless --proxy=proxy --user=user ssh user@node01
+Complete headless authentication in your local web browser:
+https://proxy.example.com/headless/<id>/approve
+// Wait for user to complete local authentication with MFA
+<user@node01> $
+```
+
+#### Headless authentication URL
+
+When the user goes to the headless authentication URL, the user will be prompted to login with MFA, if they are not logged in already. The user will then be notified of additional request details. Finally, the user is asked to verify with MFA to approve the request.
+
+Example (Exact UI/UX TBD):
+
+```
+Headless login attempt requires approval. Contact your administrator if you didn't initiate this login attempt.
+Additional details:
```

Comment:
Will we also display the `tsh` command that was used to initialize the request?

Reply:
The current design does not have any way to guarantee the command that the client will call, so it's better not to take the client provided command and display it to the user. This could give the user a false sense of security when approving a headless authentication request.